File: LongestPalindromicSubstring/longest_palindrome.py
#!/usr/bin/env python3

class Solution:
    def longestPalindrome(self, s: str) -> str:
        max_pali = ''
        max_len = 0

        for i in range(len(s)):
            l, r = i, i
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            if max_len < r - l - 1:
                max_pali = s[l+1: r]
                max_len = r - l - 1

            l, r = i, i+1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            if max_len < r - l - 1:
                max_pali = s[l+1: r]
                max_len = r - l - 1

        return max_pali

    

File: LongestPalindromicSubstring/longest_palindrome.cpp
#include <string>
#include <algorithm>
class Solution {
public:
    string longestPalindrome(string s) {
        int max_len = 0;
        std::string max_pali{};
        for ( int i = 0; i < s.size(); ++i ){
            int l{i}, r{i};
            while (0 <= l && r < s.size() && s[l] == s[r]){
                --l;
                ++r;
            }

            if (max_len < r - l - 1){
                max_len = r - l - 1;
                max_pali = s.substr(l+1, max_len);
            }

            l = i;
            r = i+1;
            while (0 <= l && r < s.size() && s[l] == s[r]){
                --l;
                ++r;
            }
            
            if (max_len < r - l - 1){
                max_len = r - l - 1;
                max_pali = s.substr(l+1, max_len);
            }
        }
        return max_pali;
    }
};

